deps/rabbit/src/rabbit_peer_discovery.erl

```diff
@@ -101,6 +114,216 @@ maybe_init() ->
+       #{domain => ?RMQLOG_DOMAIN_PEER_DISC}),
+    init_single_node(CreateClusterCallback);
+join_discovered_peers_with_retries(
+  TryNodes, NodeType, RetriesLeft, DelayInterval, CreateClusterCallback) ->
+    case find_reachable_peer_to_cluster_with(TryNodes) of
+        {ok, Node} ->
+            ?LOG_INFO(
+               "Peer discovery: Node '~ts' selected for auto-clustering",
+               [Node],
+               #{domain => ?RMQLOG_DOMAIN_PEER_DISC}),
+            create_cluster(Node, NodeType, CreateClusterCallback);
+        none ->
+            RetriesLeft1 = RetriesLeft - 1,
+            ?LOG_INFO(
+               "Peer discovery: Trying to join discovered peers failed. "
+               "Will retry after a delay of ~b ms, ~b retries left...",
+               [DelayInterval, RetriesLeft1],
+               #{domain => ?RMQLOG_DOMAIN_PEER_DISC}),
+            timer:sleep(DelayInterval),
+            join_discovered_peers_with_retries(
+              TryNodes, NodeType, RetriesLeft1, DelayInterval,
+              CreateClusterCallback)
+    end.
+
+find_reachable_peer_to_cluster_with([]) ->
+    none;
+find_reachable_peer_to_cluster_with([Node | Nodes]) when Node =/= node() ->
+    case rabbit_db_cluster:check_compatibility(Node) of
+        ok ->
+            {ok, Node};
+        Error ->
+            ?LOG_WARNING(
+               "Peer discovery: Could not auto-cluster with node ~ts: ~0p",
+               [Node, Error],
+               #{domain => ?RMQLOG_DOMAIN_PEER_DISC}),
+            find_reachable_peer_to_cluster_with(Nodes)
+    end;
+find_reachable_peer_to_cluster_with([Node | Nodes]) when Node =:= node() ->
+    find_reachable_peer_to_cluster_with(Nodes).
+
+init_single_node(CreateClusterCallback) ->
+    IsVirgin = rabbit_db:is_virgin_node(),
+    rabbit_db_cluster:ensure_feature_flags_are_in_sync([], IsVirgin),
+    CreateClusterCallback(none, disc),
+    ok.
+
+create_cluster(RemoteNode, NodeType, CreateClusterCallback) ->
+    %% We want to synchronize feature flags first before we update the cluster
+    %% membership. This is needed to ensure the local list of Mnesia tables
+    %% matches the rest of the cluster for example, in case a feature flag
+    %% adds or removes tables.
+    %%
+    %% For instance, a feature flag may remove a table (so it's gone from the
+    %% cluster). If we were to wait for that table locally before
+    %% synchronizing feature flags, we would wait forever; indeed the feature
+    %% flag being disabled before sync, `rabbit_table:definitions()' would
+    %% return the old table.
+    %%
+    %% Feature flags need to be synced before any change to Mnesia membership.
+    %% If enabling feature flags fails, Mnesia could remain in an inconsistent
+    %% state that prevents later joining the nodes.
+    IsVirgin = rabbit_db:is_virgin_node(),
+    rabbit_db_cluster:ensure_feature_flags_are_in_sync([RemoteNode], IsVirgin),
```

Comment:
This is called in `rabbit_db:pre_init`. I believe this might be a duplicated call, but not sure as there are too many clustering code paths...

Reply:
The one in `pre_init` is used when a node boots (regular boot, including during `start_app` after a `join_cluster` for instance). The one in peer discovery is used after `pre_init` if the node joins a cluster because it won't go through `pre_init` again.